### PR TITLE
Buffer behaviour fixed according to #1916

### LINKED ISSUE #1916: Buffer capacity should describe number of values, not events
Right now, `buffer(0)` is a buffer of zero events, and `buffer(1)` is a buffer of exactly one event.

I don't think this makes much sense, because it means that a 0-buffer will not replay anything, and a 1-buffer can never replay a value after it's terminated.

`RACReplaySubject`, on the other hand, treats the capacity as the number of _values_ that should be kept, and always replays termination. I think we should match that behavior with `buffer()`.

### PR DESCRIPTION
**Pull request by gavrix**
Fixes #1916.

### CODE REVIEW

File: ReactiveCocoa/Swift/SignalProducer.swift [hunk 1]
```diff
@@ -118,6 +119,9 @@ public struct SignalProducer<T, E: ErrorType> {
 			for event in events {
 				observer.put(event)
 			}
+			if let termEvent = terminationEvent {
```
Comment (jspahrsummers) [inline ReactiveCocoa/Swift/SignalProducer.swift]: You can use `let terminationEvent` here, to shadow the variable with the unwrapped version.

File: ReactiveCocoa/Swift/SignalProducer.swift [hunk 2]
```diff
@@ -111,6 +111,7 @@ public struct SignalProducer<T, E: ErrorType> {
 		lock.name = "org.reactivecocoa.ReactiveCocoa.SignalProducer.buffer"
 
 		var events: [Event<T, E>] = []
+		var terminationEvent : Event<T,E>?
```
Comment (jspahrsummers) [inline ReactiveCocoa/Swift/SignalProducer.swift]: Can you please remove the space to the left of the `:`, to match the surrounding code?

File: ReactiveCocoaTests/Swift/SignalProducerSpec.swift [hunk 3]
```diff
@@ -233,6 +233,29 @@ class SignalProducerSpec: QuickSpec {
...
 				expect(error).to(equal(TestError.Default))
 			}
+			
+			it("should always replay termination event") {
+				let (producer, sink) = SignalProducer<Int, TestError>.buffer(0)
+				var completed = false
+				sendCompleted(sink)
+				
+				producer .start(
+					completed: {
```
Comment (jspahrsummers) [inline ReactiveCocoaTests/Swift/SignalProducerSpec.swift]: Can you please reformat this to match the other code in these tests?

File: ReactiveCocoaTests/Swift/SignalProducerSpec.swift [hunk 4]
```diff
@@ -233,6 +233,29 @@ class SignalProducerSpec: QuickSpec {
...
+			it("should replay values after being terminated") {
+				let (producer, sink) = SignalProducer<Int, TestError>.buffer(1)
+				var value : Int?
+				
+				sendNext(sink, 123)
+				sendCompleted(sink)
+				producer.start(next: {val in
+					value = val
+				})
+				expect(value).to(equal(123))
```
Comment (jspahrsummers) [inline ReactiveCocoaTests/Swift/SignalProducerSpec.swift]: Can this also verify that `.Completed` was sent after the value?

File: ReactiveCocoaTests/Swift/SignalProducerSpec.swift [hunk 5]
```diff
@@ -233,6 +233,29 @@ class SignalProducerSpec: QuickSpec {
...
+				let (producer, sink) = SignalProducer<Int, TestError>.buffer(0)
+				var completed = false
+				sendCompleted(sink)
+				
+				producer .start(
+					completed: {
+					completed = true
+				})
+				expect(completed).to(beTruthy())
+			}
```
Comment (jspahrsummers) [inline ReactiveCocoaTests/Swift/SignalProducerSpec.swift]: Can you please add a blank line in between these examples?

**Comment by jspahrsummers** [comment]
This looks great, thanks! :sparkles:

I've left just a few minor notes above.

**Comment by gavrix** [comment]
@jspahrsummers I made some adjustements according to comments, let me know if there if there's something yet to be improved.

**Comment by jspahrsummers** [comment]
Looks good! :gem:

**Pull request merged**